Modernize the cookbook

tas50 opened the pull request:
The start of a modernization effort. 

- Cookstyle fixes
- Convert Vagrantfile testing to to Test Kitchen
- Add Delivery Local Mode config for runing spec/lint
- Add Github Actions for kicking off Delivery Local Mode on PRs
- Add all the modern boilerplate
- Convert the LWRPs to Custom Resources
- Remove Windows 2008 R2 support since that's EOL
- Improve logging
- Add proper name properties to each resource
- Removes the dep on the windows cookbook by requiring Chef Infra Client 14+
- Switch the resources to use powershell_out instead of the powershell_script resource so that we can get nice clean log output

tas50 committed:
Cookstyle fixes

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Format and reword the readme

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add ignore files

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Put metadata at the top of the Berkfile

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add a delivery config

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add a delivery config

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Cookstyle fixes and Github templates

Signed-off-by: Tim Smith <tsmith@chef.io>

derekgroh commented:
This looks good, do you think `Policyfiles` and `test kitchen` should be introduced as part of the modernization?

tas50 commented:
I was going to get things into Test Kitchen. I'm less worried about Berks since it's just used to depsolve the cookbooks that are needed for TK.

tas50 committed:
Remove support for Windows 2008

This is EOL now

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add a Kitchen Config

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Remove Windows 2008 support from the resources

This reduces the complexity and isn't necessary now that 2k8r2 is EOL

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Remove 2k8r2 from the readme

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Remove empty / bogus tests

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add basic inspec test

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Fully move to Test Kitchen from Vagrant

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Nuke the old vagrant file

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Use an array for windows_feature

This requires a mildly modern version of the windows cookbook (or chef where that's built in)

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Convert the LWRPs to custom resources

This makes them smaller and easier to read

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Convert to using converge_by

This gives us proper logging on converges

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add property name properties to each resource

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Require Chef Infra Client 14+ and remove the windows dep

This is a currently supported release of Chef Infra Client and it has all the windows support built in now.

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Switch the resource to powershell_out

This gives us nice clean logging and simpler code:

```
       Recipe: test_windows_dhcp::scopes
         * windows_dhcp_scope[Test] action create
         * windows_dhcp_scope[Test-Copy] action create
         * windows_dhcp_scope[Test1] action create
           - create scope Test1
         * windows_dhcp_scope[Test1] action delete
           - delete scope Test1
         * windows_dhcp_scope[Test2] action delete
       Recipe: test_windows_dhcp::leases
         * windows_dhcp_lease[Workstation3] action create
         * windows_dhcp_lease[Workstation4] action create
           - create lease Workstation4
         * windows_dhcp_lease[Workstation4] action delete
           - delete lease Workstation4
       Recipe: test_windows_dhcp::reservations
         * windows_dhcp_reservation[Workstation2] action create
           - create reservation Workstation2
         * windows_dhcp_reservation[Workstation2] action delete
           - delete reserveation Workstation2
```

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 commented:
@derekgroh pretty sure this is ready at this point

derekgroh commented on .rubocop.yml:
```diff
@@ -0,0 +1,16 @@
+
+Lint/UselessAssignment:
+  Exclude:
+    - 'Vagrantfile'
```
This can be removed

tas50 replied:
Fixed

derekgroh commented on README.md:
```diff
@@ -1,25 +1,25 @@
-windows_dhcp Cookbook
-=====================
+# windows_dhcp Cookbook
 [![Join the chat at https://gitter.im/TAMUArch/cookbook.windows_dhcp](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/TAMUArch/cookbook.windows_dhcp?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
```
Not really part of modernize, but this isn't used or monitored so I think it should be removed as well.

tas50 replied:
Fixed

derekgroh commented on metadata.rb:
```diff
@@ -3,7 +3,6 @@
-depends          'windows'
\ No newline at end of file
+supports         'windows', '>= 6.2'
+chef_version     '> 14'
```
`issues_url` and `source_url` should be added

tas50 replied:
Fixed

derekgroh reviewed (changes requested):
The data bag should be moved to the test folder as it isn't a core component of this resource cookbook and serves as an example.

A MIT `LICENSE` file should also be included to resolve all foodcritic warnings.

tas50 committed:
Updates from review

Signed-off-by: Tim Smith <tsmith@chef.io>

derekgroh commented:
LGTM

tas50 commented:
Closing due to lack of response
